hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -3802,7 +3803,12 @@ boolean internalReleaseLease(Lease lease, String src, INodesInPath iip,
             lastBlock.getBlockType());
       }
 
-      if (uc.getNumExpectedLocations() == 0 && lastBlock.getNumBytes() == 0) {
+      int minLocationsNum = 1;
+      if (lastBlock.isStriped()) {
+        minLocationsNum = ((BlockInfoStriped) lastBlock).getRealDataBlockNum();
+      }
+      if (uc.getNumExpectedLocations() < minLocationsNum &&
```

Comment:
Hi, @zhangshuyan0 ,  Also have a question here,  is `uc.getNumExpectedLocations()` always `>=` minLocationsNum?  because when invoke getAdditionalBlock rpc, the field replicas in 
 classBlockUnderConstructionFeature has been set to targets.

Reply:
First, the writing process of EC files never call `getAdditionalBlock`. Second, after a failover in NameNode, the content of `BlockUnderConstructionFeature#replicas` is totally depends on block reports (IBR or FBR).

Reply:
@zhangshuyan0 Thanks a lot. I ignored the failover condition.

Reply:
Here I think the writing process of EC files will call getAdditionalBlock and will set ReplicaUnderConstruction[] replicas , for standby name need run failover ha  will  set BlockUnderConstructionFeature#replicas is totally when block reports (IBR or FBR). ?

Reply:
> First, the writing process of EC files never call `getAdditionalBlock`. Second, after a failover in NameNode, the content of `BlockUnderConstructionFeature#replicas` is totally depends on block reports (IBR or FBR).

Sorry, I mistook `getAdditionalBlock` for `getAdditionalDatanode` just now, but the conclusion still holds because the failover occurs.

Reply:
@zhangshuyan0 Hi, shuyan. please also check below code snippet in method FSDirWriteFileOp#storeAllocatedBlock:
```java
    final BlockType blockType = pendingFile.getBlockType();
    // allocate new block, record block locations in INode.
    Block newBlock = fsn.createNewBlock(blockType);
    INodesInPath inodesInPath = INodesInPath.fromINode(pendingFile);
    saveAllocatedBlock(fsn, src, inodesInPath, newBlock, targets, blockType);

    persistNewBlock(fsn, src, pendingFile);
```
Does the BlockUnderConstructionFeature#replicas also write to editlog because it is a part of lastBlock.
Thanks a lot.

Reply:
@hfutatzhanghb `AddBlockOp` only stores blockId, numBytes and generationStamp for the last block.